src/payment/strategies/masterpass/masterpass-payment-strategy.ts

```diff
@@ -0,0 +1,153 @@
+
+        return super.deinitialize(options);
+    }
+
+    execute(payload: OrderRequestBody, options?: PaymentRequestOptions): Promise<InternalCheckoutSelectors> {
+        const order = { useStoreCredit: payload.useStoreCredit };
+
+        if (!this._paymentMethod) {
+            throw new MissingDataError(MissingDataErrorType.MissingPaymentMethod);
+        }
+
+        const gateway = this._paymentMethod.initializationData.gateway;
+        const paymentData = this._paymentMethod.initializationData.paymentData;
+
+        if (!gateway) {
+            throw new InvalidArgumentError('Unable to proceed because "paymentMethod.initializationData.gateway" argument is not provided.');
+        }
+
+        if (!paymentData) {
+            throw new InvalidArgumentError('Unable to proceed because "paymentMethod.initializationData.paymentData" argument is not provided.');
+        }
+
+        return this._store.dispatch(this._orderActionCreator.submitOrder(order, options))
+            .then(() => this._store.dispatch(this._paymentMethodActionCreator.loadPaymentMethod(gateway)))
+            .then(() => this._store.dispatch(this._paymentActionCreator.submitPayment({ methodId: gateway, paymentData })));
+    }
+
+    private _createMasterpassPayload(paymentMethod: PaymentMethod, checkout: Checkout, storeConfig: StoreConfig): MasterpassCheckoutOptions {
+        return {
+            checkoutId: paymentMethod.initializationData.checkoutId,
+            allowedCardTypes: paymentMethod.initializationData.allowedCardTypes,
+            amount: checkout.subtotal.toFixed(2),
+            currency: storeConfig.currency.code,
+            cartId: checkout.cart.id,
+            suppressShippingAddress: true,
+        };
+    }
+
+    @bind
+    private _handleWalletButtonClick(event: Event) {
+        event.preventDefault();
+
+        if (this._walletCallback) {
+            this._walletCallback();
+        }
+    }
+
+    private _hasPaymentInfo(): boolean {
+        return this._paymentMethod
+            && this._paymentMethod.initializationData
+            && this._paymentMethod.initializationData.paymentData
+            && this._paymentMethod.initializationData.paymentData.nonce;
+    }
+
+    private _masterpassClientSetup(): Promise<MasterpassCheckoutCallback> {
+        const state = this._store.getState();
+        const checkout = state.checkout.getCheckout();
+        const storeConfig = state.config.getStoreConfig();
+
+        if (!checkout) {
+            throw new MissingDataError(MissingDataErrorType.MissingCheckout);
+        }
+
+        if (!storeConfig) {
+            throw new MissingDataError(MissingDataErrorType.MissingCheckoutConfig);
+        }
+
+        if (!this._paymentMethod || !this._paymentMethod.initializationData) {
+            throw new MissingDataError(MissingDataErrorType.MissingPaymentMethod);
+        }
+
+        const payload = this._createMasterpassPayload(this._paymentMethod, checkout, storeConfig);
```

Comment:
I think you need to create the `payload` object as soon as you need it. Otherwise, you might be sending stale data to Masterpass. This is because `payload` is cached inside this closure and might be outdated by the time `_walletCallback` is called. For example, the shopper might apply a coupon code during that time. So, instead of returning a callback here, you should return the `MasterpassClient` instance instead and then use it directly inside `_handleWalletButtonClick`. i.e.:

```ts
initialize() {
    // Some code here
    return this._masterpassClientSetup()
        .then(masterpassClient) => {
            this._masterpassClient = masterpassClient;
        });
}

// And then in 
_handleWalletButtonClick() {
    // Some code here
    if (!this._masterpassClient) {
        return;
    }

    // More code here to get storeConfig, checkout data etc... or move it to `_createMasterpassPayload` so you don't have to do the preparation in this method.
    this._masterpassClient.checkout(this._createMasterpassPayload());
}
```

Reply:
@michxff Did we address this comment?